SDL main cleanup, part 1

johnnovak opened the pull request:
# Description

No functional changes (if I haven't screwed anything up), just a lift-and-shift job, basically.

Definitely review commit by commit.

I have grand plans for sdlmain, such as adding OSD support, a multi-pass shader pipeline, collapse all the CRT shaders into a single shader that will be parameterised at runtime, add brightness/contrast/saturation/white balance/etc. monitor controls, and more! I'm intending to add all that soon(ish) (definitely this year, and for 0.83.0), but I need to rein the chaos in a bit first.

Parts 2 & 3 will incorporate parts of GranMinigun's SDL & shader rework/cleanup (see https://github.com/dosbox-staging/dosbox-staging/issues/3971) and moving the OpenGL stuff into its own file.

# Manual testing

- Start up a few games
- Test various the command line params still work
- Test restarting works
- Test pause/unpause and the pause when inactive mode still works

The change has been manually tested on:

- [ ] Windows
- [x] macOS
- [x] Linux


# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/docs/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/docs/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] my change has been manually tested on Windows, macOS, and Linux.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] provided the release notes draft (for significant user-facing changes).
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

johnnovak commented on src/gui/sdlmain.cpp:
```diff
@@ -101,67 +101,89 @@ static void restore_console_encoding()
-typedef void (APIENTRYP PFNGLUNIFORM1IPROC) (GLint location, GLint v0);
-typedef void (APIENTRYP PFNGLUSEPROGRAMPROC) (GLuint program);
-typedef void (APIENTRYP PFNGLVERTEXATTRIBPOINTERPROC) (GLuint index, GLint size, GLenum type, GLboolean normalized, GLsizei stride, const GLvoid *pointer);
+typedef void(APIENTRYP PFNGLATTACHSHADERPROC)(GLuint program, GLuint shader);
```
Yeah, I was thinking the same. Most likely I'll do it in my next PR. I'll also get rid of our home-grown OpenGL proc loader and I'll use [glad](https://github.com/Dav1dde/glad) instead (in header-only mode).

FeralChild64 commented on src/gui/sdlmain.cpp:
```diff
@@ -4761,6 +4761,37 @@ static void init_logger(const CommandLineArguments* args, int argc, char* argv[]
 	loguru::init(argc, argv);
 }
 
+static void maybe_write_primary_config(const CommandLineArguments* args)
```
Unless I have a good reason, I prefer to pass a reference instead of a pointer - it's safer.

FeralChild64 commented on src/gui/sdlmain.cpp:
```diff
@@ -4792,6 +4792,52 @@ static void maybe_write_primary_config(const CommandLineArguments* args)
 }
 
+static std::optional<int> maybe_handle_command_line_output_only_actions(
+        const CommandLineArguments* args, char* argv[])
```
Again, I prefer a reference instead if pointer. And we don't need the whole `char* argv[]` inside - passing `argv[0]` (program name) should be enough.

johnnovak replied:
Good points, done both.

FeralChild64 commented:
Just a couple remarks from my side.

I have a question: what's the rationale behind removing the SDL 2.0.18 high-DPI workaround? Is it confirmed this is not needed anymore?

johnnovak commented on src/gui/sdlmain.cpp:
```diff
@@ -259,6 +259,109 @@ static void get_opengl_proc_addresses()
+	}
+}
+
+static bool LoadGLShaders(const std::string& source, GLuint* vertex, GLuint* fragment)
```
Good point, done.

johnnovak commented:
> I have a question: what's the rationale behind removing the SDL 2.0.18 high-DPI workaround? Is it confirmed this is not needed anymore?

Oops, my bad. That's still needed, I misinterpreted that version check.

Fixed that up and removed some more version checks in new commits (but kept the code intact).

johnnovak commented:
Thanks for the review and Linux testing @FeralChild64 🏅
